### PR TITLE
Feature/devel cpp aql matches

### PR DESCRIPTION
**Pull request by baslr**

### CODE REVIEW

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -2409,6 +2409,46 @@ AqlValue Functions::Document(arangodb::aql::Query* query,
...
+    retIdx = ExtractFunctionParameterValue(trx, parameters, 2).toBoolean();
+  }
+
+  AqlValueMaterializer materializer(trx);
+  VPackSlice docSlice = materializer.slice(docToFind, false);
+
+  int32_t idx = -1;
+  for (auto const& name : VPackArrayIterator(materializer.slice(exampleDocs, false))) {
+    idx++;
+    if (basics::VelocyPackHelper::VPackEqual()(name, docSlice)) {
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: better use VelocyPackHelper::compare(...) with Options being passed. 
```
VPackOptions* options = trx->transactionContextPtr()->getVPackOptions ...
```
If no Options are passed, then documents cannot be compared.

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -2409,6 +2409,46 @@ AqlValue Functions::Document(arangodb::aql::Query* query,
...
+      if (retIdx) {
+        transaction::BuilderLeaser builder(trx);
+        builder->add(VPackValue(idx));
+        return AqlValue(builder.get());
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: simply use `AqlValueHintInt(idx)` instead of Builder.

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -2409,6 +2409,46 @@ AqlValue Functions::Document(arangodb::aql::Query* query,
...
+      } else {
+        return AqlValue(arangodb::basics::VelocyPackHelper::TrueValue());
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: simply use `AqlValueHintBool(true)`

File: arangod/Aql/Functions.cpp [hunk 4]
```diff
@@ -2409,6 +2409,46 @@ AqlValue Functions::Document(arangodb::aql::Query* query,
...
+    }
+  }
+
+  if (retIdx) {
+    transaction::BuilderLeaser builder(trx);
+    builder->add(VPackValue(-1));
+    return AqlValue(builder.get());
+  }
+
+  return AqlValue(arangodb::basics::VelocyPackHelper::FalseValue());
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: simply use `AqlValueHintBool(false)`

File: arangod/Aql/Functions.cpp [hunk 5]
```diff
@@ -2409,6 +2409,46 @@ AqlValue Functions::Document(arangodb::aql::Query* query,
...
+      } else {
+        return AqlValue(arangodb::basics::VelocyPackHelper::TrueValue());
+      }
+    }
+  }
+
+  if (retIdx) {
+    transaction::BuilderLeaser builder(trx);
+    builder->add(VPackValue(-1));
+    return AqlValue(builder.get());
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: see above

**Comment by jsteemann** [comment]
The new version of `MATCHES` causes tests to fail:
```
http://c1.triagens-gmbh.zz:8080/job/arangodb-ci/job/PR-3329/5//artifact/linux-community-maintainer/03-test/singleserver-mmfiles-FAIL/shell_server_aql_3.log
```
Please fix!

**Comment by jsteemann** [comment]
I think now it's almost right.
With the exception that when comparing an object that does not have a specific property with an object that has it but with a value of `null`, then the comparison should return true.
E.g.
`{ a: null }` will match `{ }`.
`{ a: 1 }` will not match `{ }`.
`{ a: 1, b: null }` will match `{ a: 1 }`.

**Comment by baslr** [comment]
is the example on the left side?

so in arangod {something:null} is treated as it was never there? Exact the opposite as in JavaScript where you set {something:null} on purpose so its not undefined.

**Comment by jsteemann** [comment]
Please check how other parts in AQL are doing it, and how the JavaScript MATCHES implementation currently handles it. This should be considered the "reference" implementation.

**Comment by baslr** [comment]
@jsteemann 
ok the best example is 
```
let doc = {y:null, z:0}
return [
is_null(doc.x),
is_null(doc.y),
is_null(doc.z)]
```
which is `[true, true, false]`

**Comment by baslr** [comment]
matches()
if example keys value is null and docs value for examples key is null -> ok
if examples key value is null and docs value for examples key is !null -> !ok

**Pull request merged**